### DISCUSSION TITLE
TurboModule Deep Dive

### DISCUSSION THREAD

**Discussion by cortinico** [reactwg/react-native-new-architecture#2] [Deep Dive]
Hi all,
I'm opening this discussion to discuss about one of the component of the New React Native Architecture: **the TurboModule System**.

Feel free to ask your questions/ideas/doubts related to this TurboModule in this discussion.

## Relevant resources

- [Enabling TurboModule support on Android](https://reactnative.dev/docs/next/new-architecture-app-modules-android)
- [Enabling TurboModule support on iOS](https://reactnative.dev/docs/next/new-architecture-app-modules-ios)
- [Initial TurboModule proposal](https://github.com/react-native-community/discussions-and-proposals/blob/master/proposals/0002-Turbo-Modules.md)
- https://github.com/react-native-community/discussions-and-proposals/issues/40
- https://github.com/react-native-community/discussions-and-proposals/pull/11

**Comment by jacobp100** [comment]
Something I can't quite find the answer to is the status of the bridge. I've heard the term 'bridgless' thrown about, but turbo modules still have a bridge.

I'm currently the only person maintaining a web-workers-like module, [which uses the bridge](https://github.com/jacobp100/react-native-threads/blob/master/ios/ThreadManager.m). I want to take a stab at migrating this over to the new architecture, but wanted to know if this approach is going to be ok going forward.

**Reply by cipolleschi** [reply]
Hi @jacobp100, thanks for your question.

Currently yes, TurboModule still require the bridge to be initialized for backward compatibility. 
When we use the TurboModule with the new architecture enabled, what happens is that the communication does not go through the bridge, but it uses `JSI` to share memory between the JS part and the native part. 

In practice, JS holds references to C++ objects, and it can call methods on those objects directly, instead of serializing the required data in JSON and send them through the bridge.

I hope that this answer your question, feel free to ask more if you need. We are also working on improving the documentation, so more information will arrive soon.

PS: sorry for the late reply!

**Reply by jacobp100** [reply]
Thanks for your reply - that makes a lot of sense!

**Reply by cortinico** [reply]
To add to @cipolleschi 's answer, I can totally recommend this talk from Emily Janzer if you want to learn more about bridgeless mode & the New React Native Architecture:
https://www.youtube.com/watch?v=52El0EUI6D0

**Reply by ClaudiuHBann** [reply]
> Hi @jacobp100, thanks for your question.
> 
> Currently yes, TurboModule still require the bridge to be initialized for backward compatibility. When we use the TurboModule with the new architecture enabled, what happens is that the communication does not go through the bridge, but it uses `JSI` to share memory between the JS part and the native part.
> 
> In practice, JS holds references to C++ objects, and it can call methods on those objects directly, instead of serializing the required data in JSON and send them through the bridge.
> 
> I hope that this answer your question, feel free to ask more if you need. We are also working on improving the documentation, so more information will arrive soon.
> 
> PS: sorry for the late reply!

What are the `Bridging` template specializations for? Just to map the properties of an c++ complex object to javascript?
For example the `Bridging` specialization for `std::string` will copy the data into a `jsi::String`:
![image](https://github.com/user-attachments/assets/029af4ae-dfa7-47b5-aae9-a95958da7a70)

If i have a `std::map<std::string, std::string>` this will be pretty expensive to "reference" it into the javascript runtime.

Can I move my C++ objects into the JSI objects? if not, can I at least have a view over them?

**Comment by tido64** [comment]
I'd like to know (or at least find out) what needs to be done in order to support autolinking TurboModules in the CLI. From what I can see, there's a lot of boilerplate code, and most of it (if not all) look like they could be generated. I also think that we should have this in place sooner than later.

cc @thymikee, @jsamr

**Reply by cipolleschi** [reply]
Hi @tido64, thanks for your question. Can you point out the boilerplate code, please?

I can talk for iOS, while @cortinico may answer you about Android. In iOS, to enable autolinking, we need:
1. a `podspec` file for the module;
2. a `xcodeproj` situated in the `ios` folder.

If those two requirements are satisfied, the `pod install` step will be able to autolink the module.

I agree that we can improve the situation: I think that the requirement for the `xcodeproj` should be dropped and that many files can be generated automatically. 

I hope this helps, let me know if you have more questions on the iOS part.

PS: sorry for the late reply!

**Reply by thymikee** [reply]
@cipolleschi so it looks like nothing changes really and autolinking should "just work" for already supported modules?  FYI since v8 of the RN CLI (currently in alpha, targetting RN 0.69) we only take podspec and Podfile into account: https://github.com/react-native-community/cli/commit/25eec7c695f09aea0ace7c0b591844fe8828ccc5 so no xcodeproj is necessary (as it can be generated).  cc @grabbou

**Reply by cipolleschi** [reply]
> autolinking should "just work" for already supported modules?

Yes, that's right. As stated [here](https://reactnative.dev/docs/next/new-architecture-library-intro#make-sure-autolinking-is-enabled), for iOS we just need a podspec.

> since v8 of the RN CLI (currently in alpha, targetting RN 0.69) we only take podspec and Podfile into account: https://github.com/react-native-community/cli/commit/25eec7c695f09aea0ace7c0b591844fe8828ccc5 so no xcodeproj is necessary (as it can be generated)

Amazing, thank you for this.

**Reply by tido64** [reply]
I haven't had time to sit down and play with this yet, from just reading the docs:

- On iOS, if the setup does not require additional changes when adding community modules, I think it looks fine.
- On Android, [step 4](https://reactnative.dev/docs/next/new-architecture-app-modules-android#4-extend-the-getpackages-from-your-reactnativehost-to-use-the-turbomodule) requires changes for each additional TurboModule if I understand correctly. `@react-native-community/cli` does this today for native modules, but will have to add support for TurboModules. It also looks like you need to add an entry to the file made in [step 5](https://reactnative.dev/docs/next/new-architecture-app-modules-android#5-c-provide-a-native-implementation-for-the-methods-in-your-turbomoduledelegate-class). Does this sound right? If so, both of these should be generated by `@react-native-community/cli` as well.

**Reply by cortinico** [reply]
Hey all, sorry for the late reply on Android. I was testing this and need to further test it to fully understand if there are other things to address on Android. Yet, I can share what I believe is needed so far.

Sadly, autolinking on Android will require a bit more work to fully work with the New Arch. Specifically the CLI codegen will need to be extended to generate a couple of more files.

In the current status (i.e. with the upcoming 0.68.0 and the new app template), users will still be able to link against 3rd party libs that are exposing TM or Fabric components, but they will have to do some manual steps.

First, autolinking will have to rely on the library name specified in the `build.gradle` file. For the sake of the examples below, let's call this library name `moka`. 

## Changes to the Make/CMake files

Users will have to import the generate `.mk` files of the library inside the `Android.mk` of their application ([this file](https://github.com/facebook/react-native/blob/main/template/android/app/src/main/jni/Android.mk))

Namely those changes are needed:

```diff
 include $(REACT_ANDROID_DIR)/Android-prebuilt.mk
 
 # If you wish to add a custom TurboModule or Fabric component in your app you
 # will have to include the following autogenerated makefile.
 # include $(GENERATED_SRC_DIR)/codegen/jni/Android.mk

+# Includes the .mk file for library moka
+include $(NODE_MODULES_DIR)/moka/android/app/build/generated/source/codegen/jni/Android.mk

 [...]

 LOCAL_SHARED_LIBRARIES := \
   libfabricjni \
   libfbjni \
   libfolly_runtime \
   libglog \
   libjsi \
   libreact_codegen_rncore \
+  libreact_codegen_moka \
   libreact_debug \
   libreact_nativemodule_core \
   libreact_render_componentregistry \
   libreact_render_core \
   libreact_render_debug \
   libreact_render_graphics \
   librrc_view \
   libruntimeexecutor \
   libturbomodulejsijni \
   libyoga
```

Please note that the `NODE_MODULES_DIR` env variable would have to be set in the build.gradle of the template.

Agree that the include is not great. We could potentially have a `$(call import-codegen-module,moka)` which will make for a easier and more maintainable syntax.

Moreover, we're currently updating the `.mk` files to CMake files, which will make the setup easier.

To work with autolinking, we would have to adapt this setup to work with a generated `Android.mk`/`CMakeLists.txt` file which contains all the imports of the discovered library. 

## Changes to the Application Module Provider

The `MainApplicationModuleProvider.cpp` file ([this one](https://github.com/facebook/react-native/blob/main/template/android/app/src/main/jni/MainApplicationModuleProvider.cpp)) needs to be updated to include all the autolinked modules. This is necessary in order for TMs to be discovered.

Namely a user would have to do this change to include the `moka` library:

```diff
 #include "MainApplicationModuleProvider.h"

 #include <rncore.h>
+#include <moka.h>

 namespace facebook {
 namespace react {

 std::shared_ptr<TurboModule> MainApplicationModuleProvider(
     const std::string moduleName,
     const JavaTurboModule::InitParams &params) {
   // Here you can provide your own module provider for TurboModules coming from
   // either your application or from external libraries. The approach to follow
   // is similar to the following (for a library called `samplelibrary`:
   //
   // auto module = samplelibrary_ModuleProvider(moduleName, params);
   // if (module != nullptr) {
   //    return module;
   // }
   // return rncore_ModuleProvider(moduleName, params);
+  auto module = moka_ModuleProvider(moduleName, params);
+  if(module != nullptr) {
+      return module;
+  }
   return rncore_ModuleProvider(moduleName, params);
 }
```

Therefore we probably need a `autolinked_ModuleProvider` that checks for all the autolinked modules. Those files are going to be a .h/.cpp file can be codegenned and in this if-then-else statement in a cascading style (similary to `getPackages` on the main application class). 

## Changes to the Main Components Registry

The `MainComponentsRegistry.cpp` file ([this one](https://github.com/facebook/react-native/blob/main/template/android/app/src/main/jni/MainComponentsRegistry.cpp)) needs to be updated to include all the autolinked components. This is necessary in order for external Fabric componetns to be discovered.


Namely a user would have to do this change to include the `moka` library:

```diff
 #include "MainComponentsRegistry.h"

 #include <CoreComponentsRegistry.h>
 #include <fbjni/fbjni.h>
 #include <react/renderer/componentregistry/ComponentDescriptorProviderRegistry.h>
+#include <react/renderer/components/moka/ComponentDescriptors.h>
 #include <react/renderer/components/rncore/ComponentDescriptors.h>

 namespace facebook {
 namespace react {

 MainComponentsRegistry::MainComponentsRegistry(ComponentFactory *delegate) {}

 std::shared_ptr<ComponentDescriptorProviderRegistry const>
 MainComponentsRegistry::sharedProviderRegistry() {
   auto providerRegistry = CoreComponentsRegistry::sharedProviderRegistry();

   // Custom Fabric Components go here. You can register custom
   // components coming from your App or from 3rd party libraries here.
   //
   // providerRegistry->add(concreteComponentDescriptorProvider<
   //        AocViewerComponentDescriptor>());
+  providerRegistry->add(concreteComponentDescriptorProvider<
+          MokaComponentDescriptor>());
   return providerRegistry;
 }
```

This is a bit trickier as `react/renderer/components/moka/ComponentDescriptors.h` might contain more than one generated `ComponentDescriptors` (e.g. a `CoffeeComponentDescriptors`, `CappuccinoComponentDescriptors` or more).

Again here we would need a `AutoLinkedComponentDescriptors` which exposes a list of discovered `ComponentDescriptors` that can be all added in this function here.

-----

I hope this helps provide a bit of insights on what is needed inside the CLI to make this setup smoother 👍 Let me know if you have further questions

**Reply by tido64** [reply]
Thanks for the detailed explanation. This really helps. I guess my next question becomes: Who should work on this? Is it Meta? Callstack? Others? I feel like if we want people to adopt the new architecture, this needs to be in place.

**Reply by thymikee** [reply]
@tido64 we've moved the discussion about implementing this in CLI here: https://github.com/react-native-community/cli/issues/1596. As Callstack representative I can't promise focusing on this at this very moment (we have some plans on improving maintenance of RN CLI in progress), but I'm happy to merge PRs and release.

**Reply by cortinico** [reply]
> I feel like if we want people to adopt the new architecture, this needs to be in place.

Totally agree. So far our (Meta) focus has been to ship the minimum amount of infrastructure needed to make the New Architecture available to use. We also spent significant amount of time making sure that 3rd party libs have the necessary infrastructure needed to migrate their lib.

Autolinking is a crucial piece of infrastructure as well that we'd love to have in place asap. Currently users can still use New Architecture but will have to manually link their libraries (definitely not ideal).

If someone from the CLI, Callstack or the Community is willing to take a stance on this, I'm happy to support it.

**Reply by thymikee** [reply]
Ok, Callstack will take care of this. I may reach out to you @cortinico to challenge some ideas :)

**Reply by cortinico** [reply]
Awesome, looking forward to it 👍

**Reply by thymikee** [reply]
Draft PR: https://github.com/react-native-community/cli/pull/1603

**Reply by cortinico** [reply]
Awesome, thanks for sendign this over

**Reply by shubham-ghanghotia** [reply]
> Hi @tido64, thanks for your question. Can you point out the boilerplate code, please?
> 
> I can talk for iOS, while @cortinico may answer you about Android. In iOS, to enable autolinking, we need:
> 
> 1. a `podspec` file for the module;
> 2. a `xcodeproj` situated in the `ios` folder.
> 
> If those two requirements are satisfied, the `pod install` step will be able to autolink the module.
> 
> I agree that we can improve the situation: I think that the requirement for the `xcodeproj` should be dropped and that many files can be generated automatically.
> 
> I hope this helps, let me know if you have more questions on the iOS part.
> 
> PS: sorry for the late reply!


That's great! 
@cipolleschi  How can I generate the .xcodeproj file in iOS for the library created using react-native-create-library?

**Reply by cipolleschi** [reply]
You can use Xcode to generate a new project and then add all the Swift/Objective-C/Objective-C++/C/C++ files you need.

**Comment by DanielFrTB** [comment]
Hi @cortinico ,
Does TurboModule support Running on the main thread?
on the [old arch](https://reactnative.dev/docs/native-modules-ios#threading):
`- (dispatch_queue_t)methodQueue { return dispatch_get_main_queue(); }`

**Reply by cipolleschi** [reply]
`RCTTurboModule` doesn't have the same capability of specifying a queue and have every methods running on that queue, but if you implement the method `methodQueue`, we are keeping honoring it for backward compatibility.

In general, with TM, you should be able to handle the execution queue as you desire.

But what are you trying to do? Why you need to run a TM in the main queue?

**Reply by DanielFrTB** [reply]
@cipolleschi 
We are trying to create a UIView on TurboModules and pass its instance to RCTViewComponentView.

**Reply by cipolleschi** [reply]
In the New Architecture, that's not supported nor advised. 
TurboModule system and Fabric are meant to talk to each other through JS. 
You should not try to create a View in a Turbomodule: that's also why Fabric does not need ViewManagers (which are native modules) to run in bridgeless mode.

We want to enforce a cleaner separation between Modules and Views because this architecture solves a classes of bugs we used to have in the Old Architecture due to Modules tinkering with the Views. 

If you can explain a little bit more what you want to achieve (why you need to create a UIView in a TM and pass it to a RCTViewComponentView?), we can brainstorm together what's the best approach.
